Signum.Engine.Extensions/Tree/TreeLogic.cs

```diff
@@ -184,28 +184,46 @@ private static SqlHierarchyId GetNewPosition<T>(MoveTreeModel model, TreeEntity
             where T : TreeEntity
         {
             var newParentRoute = model.NewParent == null ? SqlHierarchyId.GetRoot() : model.NewParent.InDB(a => a.Route);
-            
+
             if (newParentRoute.IsDescendantOf(entity.Route))
                 throw new Exception(TreeMessage.ImpossibleToMove0InsideOf1.NiceToString(entity, model.NewParent));
 
             if (model.InsertPlace == InsertPlace.FirstNode)
                 return newParentRoute.GetDescendant(SqlHierarchyId.Null, FirstChild<T>(newParentRoute));
 
-            if(model.InsertPlace == InsertPlace.LastNode)
+            if (model.InsertPlace == InsertPlace.LastNode)
                 return newParentRoute.GetDescendant(LastChild<T>(newParentRoute), SqlHierarchyId.Null);
 
             var newSiblingRoute = model.Sibling!.InDB(a => a.Route);
 
-            if (!newSiblingRoute.IsDescendantOf(newParentRoute) || 
-                newSiblingRoute.GetLevel() != newParentRoute.GetLevel() + 1 || 
+            if (!newSiblingRoute.IsDescendantOf(newParentRoute) ||
+                newSiblingRoute.GetLevel() != newParentRoute.GetLevel() + 1 ||
                 newSiblingRoute == entity.Route)
                 throw new Exception(TreeMessage.ImpossibleToMove01Of2.NiceToString(entity, model.InsertPlace.NiceToString(), model.NewParent));
 
             if (model.InsertPlace == InsertPlace.After)
-                return newParentRoute.GetDescendant(newSiblingRoute, Next<T>(newSiblingRoute));
-            
+            {
+                var next = Next<T>(newSiblingRoute);
+                //var newHid = newParentRoute.GetDescendant(newSiblingRoute, next);
+
+                //if (newHid == newSiblingRoute|| newHid == next)
+                 var   newHid = Database.Query<T>().Where(e=>(bool)(e.Route== newParentRoute)).Select(e => e.Route.GetDescendant(newSiblingRoute, next)).SingleEx();
+
+                return newHid;
+            }
+
+
             if (model.InsertPlace == InsertPlace.Before)

```

Comment:
Hi @MehdyKarimpour, @AlejandroCano says that  GetDescentant implementation is buggy in-memory, so this change does it in the database. Does it work for you?

Reply:
I have fixed the GetDescendants implementations https://github.com/dotMorten/Microsoft.SqlServer.Types/pull/38,

let's wait till the change is merged and the nuget package published